apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/AbstractGrpcRdaSource.java

```diff
@@ -218,6 +223,21 @@ protected int submitBatchToSink(
     return processed;
   }
 
+  /**
+   * Checks if the given api version string represents an RDA Version that this job is allowed to
+   * ingest data from.
+   *
+   * @param apiVersionString A string representing the RDA API version being checked.
+   * @throws IllegalStateException If the apiVersionString represents an API version that can't be
+   *     ingested.
+   */
+  protected void checkApiVersion(String apiVersionString) {
+    if (!rdaVersion.allows(apiVersionString)) {
+      throw new IllegalStateException(
+          String.format("Can not ingest data from API running version '%s'", apiVersionString));
```

Comment:
Could this not be handled with normal control flow instead of an exception? It seems like this is a standard validation and not really an "exception" case.

Reply:
Having the RDA API version not match the desired range is not something we currently expect to happen frequently, which arguably makes it an exceptional circumstance, and one that we want everything to stop for.

The higher logic is processing claims and returning a processing result, which can contain a processed claims count or an exception, which returning an exception vs throwing one wouldn't be much of a difference (since the logic later throws that exception).

We could add a new code path that handles it without exceptions, but that would likely mean adding logic to several layers on the way back up, which arguably greatly increases complexity.

Reply:
My $.02 would be that server incompatibility is a very legitimate exception condition.  Very much like data transformation errors are.  We have explicit checks for both cases but we expect them to happen very rarely and use exceptions to make the code cleaner and easier to understand.  Similar case can be made for `Integer.parseInt()` throwing rather than returning an error code.

I disagree that fatal errors throwing exceptions makes the code harder to understand.

Now if we were doing functional programming here and using error monads that would be a different story.  But then the language would provide tools to facilitate that.

Reply:
How would one define "ability for the code that's processing to be able to handle the error scenario with the context it has"?  If the statement here is that an event that should cause the entire pipeline job to halt is "recoverable", then what wouldn't be a recoverable?

Reply:
My concern was more the graceful handling of the error and being able to return something meaningful to the caller, not necessarily the recoverability of it. I agree that this is a fatal error and isnt recoverable to the pipeline; my only quibble was making sure the way the failure handled continues being understandable to future maintainers.

This is mostly stemming from my real headache of currently getting pipeline to exit gracefully when it fails; it uses exceptions often for its failure states but due to layered complexity over time the exception never makes it to the place the writer thought it would and doesnt do what was probably intended. Over time since exceptions invisibly bubble up to a place the flow is not easy for someone to understand/notice when adding new layers of functionality.

Anyway, not going to necessarily die on this hill or block the PR. Just pointing out some issues I'm seeing today to prevent possible issues in the future.